### DISCUSSION TITLE
Using middleware in order to implement business logic

### DISCUSSION THREAD

**Discussion by sehmaschine** [prisma/prisma#4932] [Q&A] [answered]
I'm just trying to figure out what´s the best option to implement DB updates based on certain triggers (e.g. a creating or updating a record). For example, a user creates a new record for tableA. Depending on the values of two fields in tableA we might need to create an additional record in tableB. We could implement this as part of the API, but the record for tableA could also be created based on a background task (which does not call the API). So I thought that using the middleware is appropriate for that usecase. I'm only hesitating because of the possible performance implications. 

Is using a middleware the best option here? Or would you suggest something else?
And ... if using a middlware is appropriate, would you use _one_ middleware containing all methods or a separate middleware for each method?

Coming from Python/Django: we either overwrite the models save method or use signals for that kind of behaviour. So what I'm looking for is the Prisma alternative to either of that.

**Comment by ryands17** [comment, marked as answer]
Hey @sehmaschine 👋 

Yes you can totally use Prisma's [Middleware](https://www.prisma.io/docs/concepts/components/prisma-client/middleware) for this. 

> I'm only hesitating because of the possible performance implications.

There wouldn't be any implications as long as you add a condition in the middleware that only fires if Model `A` is being called upon.

> would you use one middleware containing all methods or a separate middleware for each method?

You can go about this any way actually. If you have a lot of conditions, you can split those in different middleware by just adding another `prisma.use`.

Let me know if you have any other queries :)

**Reply by JakubKoralewski** [reply]
I'm not currently using raw SQL queries, but if I would: is there a way to trigger the middleware if running raw queries? I mean maybe somehow let Prisma know that the sql query is responsible for *this model* so then the middleware could filter events correctly? In ideal world setting the type annotation with the given model as return value should trigger that, but unfortunately I don't think that can work, right?

**Reply by sehmaschine** [reply]
Thanks @ryands17. That mainly solves one of my issues. Following up on that: It is currently not possible to deal with Postgres Triggers/Functions/Views with Prisma, right? For example, I'm not able to declare a "virtual column" (which is actually a function) with the schema, correct? I already tried (adding the col manually, which is fine for me) but it didn't work. 

And I'm also interested in what @JakubKoralewski is asking. We (unfortunately) need a lot of RAW SQL queries right now in order to bypass some of Prismas limitations. But then, the middleware is not being called I guess.

**Reply by ryands17** [reply]
@JakubKoralewski 👋 
The middleware will run surely, but it won't be able to identify the model so you would need to write the model finding logic yourself.

For eg. if I'm executing a raw query via `prisma.$queryRaw`, it will give me the following output:

![image](https://user-images.githubusercontent.com/19697099/104170259-dab94f00-5426-11eb-9d39-6286fe0f3905.png)

Here my model and table name is `User`. So you would need to write some logic to figure the model out in this case.


@sehmaschine As for triggers/views, currently Prisma schema doesn't support those yet. You can follow these issues for more progress on those:

- https://github.com/prisma/prisma/issues/3383
- https://github.com/prisma/prisma/issues/678

You can add your views and functions via Prisma Migrate, but it won't be visible in your `schema.prisma` and you would need to use [raw queries](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access) to fetch data from those.

Let me know if that answers your question.

**Reply by sehmaschine** [reply]
Thanks @ryands17. My main question is answered.

**Reply by JakubKoralewski** [reply]
@ryands17 Thanks for the detailed answer, it also answers my question, but I have a highly speculative question though. If I were to take a very esoteric approach, and [somehow parse the typescript myself for some reason](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API) to assign to the `model` property [the type given during the call to `$queryRaw`](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#typing-queryraw-results), or more simply I would just like to annotate the model by running something like `prisma.$queryRaw('my sql', {model: User})` -- is it possible to attach metadata to a prisma call like that for the middleware to consume?. 

So basically I guess what I'm asking is since the `$queryRaw` type signature is `$queryRaw<T = any>(query: string | TemplateStringsArray, ...values: any[]): Promise<T>;`, are the `...values` available for middlewares?

**Reply by ryands17** [reply]
I don't think that's possible. You can only attach variables you need to interpolate i.e. where you have `$` type variables.

**Reply by JakubKoralewski** [reply]
Thanks anyway.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
